File: wallet/src/main/java/com/worldventures/wallet/ui/settings/help/feedback/impl/SendFeedbackPresenterImpl.java
package com.worldventures.wallet.ui.settings.help.feedback.impl;


import com.worldventures.core.modules.infopages.service.FeedbackInteractor;
import com.worldventures.core.modules.picker.service.MediaPickerInteractor;
import com.worldventures.wallet.service.command.settings.WalletSettingsInteractor;
import com.worldventures.wallet.service.command.settings.help.CustomerSupportFeedbackCommand;
import com.worldventures.wallet.service.command.settings.help.SendWalletFeedbackCommand;
import com.worldventures.wallet.service.command.settings.help.SmartCardFeedbackCommand;
import com.worldventures.wallet.ui.common.base.WalletDeviceConnectionDelegate;
import com.worldventures.wallet.ui.common.navigation.Navigator;
import com.worldventures.wallet.ui.settings.help.feedback.FeedbackType;
import com.worldventures.wallet.ui.settings.help.feedback.SendFeedbackPresenter;
import com.worldventures.wallet.ui.settings.help.feedback.SendFeedbackScreen;
import com.worldventures.wallet.ui.settings.help.feedback.base.impl.BaseFeedbackPresenterImpl;

public class SendFeedbackPresenterImpl extends BaseFeedbackPresenterImpl<SendFeedbackScreen> implements SendFeedbackPresenter {

   public SendFeedbackPresenterImpl(Navigator navigator, WalletDeviceConnectionDelegate deviceConnectionDelegate,
         FeedbackInteractor feedbackInteractor,
         WalletSettingsInteractor walletSettingsInteractor, MediaPickerInteractor mediaPickerInteractor) {
      super(navigator, deviceConnectionDelegate, feedbackInteractor, walletSettingsInteractor,
            mediaPickerInteractor);
   }

   @Override
   public void attachView(SendFeedbackScreen view) {
      super.attachView(view);
      observeFormValidation();
      final FeedbackType feedbackType = getView().getFeedbackType();
      getView().applyFeedbackType(feedbackType);
   }

   private void observeFormValidation() {
      getView().getTextFeedbackMessage().map(this::validForm)
            .compose(getView().bindUntilDetach())
            .subscribe(enable -> getView().changeActionSendMenuItemEnabled(enable));
   }

   private boolean validForm(CharSequence textMessage) {
      final boolean messageIsEmpty = textMessage.toString().trim().isEmpty();
      return !messageIsEmpty && getAttachmentsManager().getFailedOrPendingAttachmentsCount() <= 0;
   }

   @Override
   public void sendFeedback(String text) {
      getView().changeActionSendMenuItemEnabled(false);

      sendFeedbackCommand(getView().getFeedbackType() == FeedbackType.SmartCardFeedback
            ? new SmartCardFeedbackCommand(text, getImagesAttachments())
            : new CustomerSupportFeedbackCommand(text, getImagesAttachments()));
   }

   @Override
   public void handleSuccessSentFeedback() {
      clearAttachments();
      goBack();
   }

   @Override
   protected void handleFailSentFeedback(SendWalletFeedbackCommand command, Throwable throwable) {
      getView().changeActionSendMenuItemEnabled(true);
   }

   @Override
   public void goBack() {
      getNavigator().goBack();
   }
}

File: wallet/src/main/java/com/worldventures/wallet/ui/settings/help/feedback/regular/impl/SendFeedbackPresenterImpl.kt
package com.worldventures.wallet.ui.settings.help.feedback.regular.impl

import com.worldventures.wallet.service.command.settings.WalletSettingsInteractor
import com.worldventures.wallet.service.command.settings.help.CustomerSupportFeedbackCommand
import com.worldventures.wallet.service.command.settings.help.SendWalletFeedbackCommand
import com.worldventures.wallet.service.command.settings.help.SmartCardFeedbackCommand
import com.worldventures.wallet.ui.common.base.WalletDeviceConnectionDelegate
import com.worldventures.wallet.ui.common.navigation.Navigator
import com.worldventures.wallet.ui.settings.help.feedback.base.impl.BaseFeedbackPresenterImpl
import com.worldventures.wallet.ui.settings.help.feedback.base.impl.FeedbackAttachmentsPresenterDelegate
import com.worldventures.wallet.ui.settings.help.feedback.regular.FeedbackType
import com.worldventures.wallet.ui.settings.help.feedback.regular.SendFeedbackPresenter
import com.worldventures.wallet.ui.settings.help.feedback.regular.SendFeedbackScreen

class SendFeedbackPresenterImpl(navigator: Navigator,
                                deviceConnectionDelegate: WalletDeviceConnectionDelegate,
                                attachmentDelegate: FeedbackAttachmentsPresenterDelegate,
                                walletSettingsInteractor: WalletSettingsInteractor)
   : BaseFeedbackPresenterImpl<SendFeedbackScreen>(navigator, deviceConnectionDelegate, attachmentDelegate, walletSettingsInteractor),
      SendFeedbackPresenter {

   override fun sendFeedback() {
      view!!.changeActionSendMenuItemEnabled(false)
//
      sendFeedbackCommand(
            if (view!!.feedbackType == FeedbackType.SmartCardFeedback)
               SmartCardFeedbackCommand(view!!.feedbackMessage, attachmentDelegate.imagesAttachments)
            else
               CustomerSupportFeedbackCommand(view!!.feedbackMessage, attachmentDelegate.imagesAttachments)
      )
   }

   override fun attachView(view: SendFeedbackScreen) {
      super.attachView(view)
      observeFormValidation()

      val feedbackType = view.feedbackType
      view.applyFeedbackType(feedbackType)
   }

   private fun observeFormValidation() {
      view!!.textFeedbackMessage
            .map { it.isNotBlank() }
            .startWith(false)
            .compose(view!!.bindUntilDetach())
            .subscribe { enable ->
               view!!.changeActionSendMenuItemEnabled(enable && !attachmentDelegate.hasFailedOrPendingAttachments)
            }
   }

   public override fun handleSuccessSentFeedback() {
      attachmentDelegate.clearAttachments()
      goBack()
   }

   override fun handleFailSentFeedback(command: SendWalletFeedbackCommand<*>, throwable: Throwable) {
      view!!.changeActionSendMenuItemEnabled(true)
   }

   override fun goBack() {
      attachmentDelegate.clearAttachments()
      navigator.goBack()
   }
}
